mapswipe_workers/mapswipe_workers/firebase_to_postgres/transfer_results.py

```diff
@@ -55,24 +42,92 @@ def transfer(current_results):
+        before the new value is successfully saved,
+        the update function is called again with the new current value,
+        and the write will be retried."
 
-        logger.info(f"{project_id}: Start transfering results")
+    Using Firebase transaction on the group level
+    has turned out to be too slow when using "normal" queries,
+    e.g. without using threading. Threading should be avoided here
+    as well to not run into unforeseen errors.
 
-        results_ref = fb_db.reference(f"v2/results/{project_id}")
+    For more details see issue #478.
+    """
+    # TODO: which exceptions can happen here?
+
+    logger.info(f"{project_id}: Start transfer results")
+    fb_db = auth.firebaseDB()
+    results_ref = fb_db.reference(f"v2/results/{project_id}")
+    results = results_ref.get()
+
+    if results is None:
+        logger.info(f"{project_id}: No results in Firebase")
+    else:
+        # First we check for new users in Firebase.
+        # The user_id is used as a key in the postgres database for the results
+        # and thus users need to be inserted before results get inserted.
+        results_user_id_list = get_user_ids_from_results(results)
+        update_data.update_user_data(results_user_id_list)
+
+        # Results are dumped into an in-memory file.
+        # This allows us to use the COPY statement to insert many
+        # results at relatively high speed.
+        results_file = results_to_file(results, project_id)
         truncate_temp_results()
+        save_results_to_postgres(results_file)
 
-        try:
-            results_ref.transaction(transfer)
-            logger.info(f"{project_id}: Transfered results to postgres")
-        except fb_db.TransactionAbortedError:
-            logger.exception(
-                f"{project_id}: Firebase transaction for "
-                f"transfering results failed to commit"
-            )
-            sentry.capture_exception()
+        # It is important here that we first insert results into postgres
+        # and then delete these results from Firebase.
+        # In case something goes wrong during the insert, results in Firebase
+        # will not get deleted.
+        delete_results_from_firebase(project_id, results)
+        del fb_db
+        logger.info(f"{project_id}: Transferred results to postgres")
+
+
+def delete_results_from_firebase(project_id, results):
+    """Delete results from Firebase using update function.
+
+    We use the update method of firebase instead of delete.
+    Update allows to delete items at multiple locations at the same time
+    and is much faster.
+    """
+
+    fb_db = auth.firebaseDB()
+
+    # we will use a multi-location update to delete the entries
+    # therefore we create a dict with the items we want to delete
+    data = {}
+    for group_id, users in results.items():
+        for user_id, result in users.items():
+            key = f"{group_id}/{user_id}"
+            data[key] = None
```

Comment:
Does setting the reference (`fb_db.reference(f"v2/results/{project_id}/")`) with a dict containing this key (`"{group_id}/{user_id}"`) work as expected?

I would build a nested dict as follows:
```python
    for group_id, users in results.items():
        data[group_id] = {}
        for user_id in users.keys():        
             data[group_id][user_is] = None
```

But it is a while since I worked with Firebase.

Reply:
Hey, @Matthias-Schaub 
it works. So basically the first part in that dict is the path. That's why it looks a bit strange.

This is also covered by the tests, in which we check that the results are actually deleted from firebase.